src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java

```diff
@@ -573,21 +573,36 @@ public Worker(
         this.maxGetRecordsThreadPool = maxGetRecordsThreadPool;
         this.workerStateChangeListener = workerStateChangeListener;
         workerStateChangeListener.onWorkerStateChange(WorkerStateChangeListener.WorkerState.CREATED);
-        this.shardSyncStrategy = createShardSyncStrategy(config.getShardSyncStrategyType());
-        LOG.info(String.format("Shard sync strategy determined as %s.", shardSyncStrategy.getStrategyType().toString()));
-        this.leaderDecider = leaderDecider != null ? leaderDecider : createLeaderDecider();
-        this.leaderElectedPeriodicShardSyncManager = periodicShardSyncManager != null ? periodicShardSyncManager
-                : createPeriodicShardSyncManager();
+        createShardSyncStrategy(config.getShardSyncStrategyType(), leaderDecider, periodicShardSyncManager);
     }
 
-    private ShardSyncStrategy createShardSyncStrategy(ShardSyncStrategyType strategyType) {
+    /**
+     * Create shard sync strategy and corresponding {@link LeaderDecider} based on provided configs. PERIODIC
+     * {@link ShardSyncStrategyType} honors custom leaderDeciders for leader election strategy. All other
+     * {@link ShardSyncStrategyType}s permit only a default single-leader strategy.
+     */
+    private void createShardSyncStrategy(ShardSyncStrategyType strategyType,
+                                         LeaderDecider leaderDecider,
+                                         PeriodicShardSyncManager periodicShardSyncManager) {
         switch (strategyType) {
             case PERIODIC:
-               return createPeriodicShardSyncStrategy();
+                this.leaderDecider = leaderDecider;
```

Comment:
For PERIODIC strategyType, can leaderDecider be null? I think we also need to check null and create the leaderDecider if null

Reply:
leaderDecider should never be null. For the Builder initialization this is covered. But you're right that this could throw NPEs if the regular constructors are used. 

This is actually a gap in the current implementation as well - good catch! Will add the null check.